Sources/Version.swift

```diff
@@ -1,39 +1,131 @@
+            throw DecodingError("Unable to decode version")
+        }
 
         let components = string.components(separatedBy: ".")
-        guard components.count >= 1, let major = Int(components[0]) else {
-            return nil
+        let integerComponents = try components.map { component -> UInt in
+            if let int = UInt(component) {
+                return int
+            }
+
+            throw DecodingError("Unable to parse version component into integer")
         }
 
-        let minor = components.count >= 2 ? Int(components[1]) : nil
-        let patch = components.count >= 3 ? Int(components[2]) : nil
-        return Version(major: major, minor: minor, patch: patch)
+        guard let lastComponent = integerComponents.last else {
+            throw DecodingError("Version components were empty")
+        }
+
+        let version = integerComponents.dropLast().reversed().reduce(Version.version(lastComponent))
+        { subversion, component in
+            return Version.subversion(component, subversion)
+        }
+
+        self = version
     }
 
-    func transformToJSON(_ value: Version?) -> String? {
-        if let value = value {
-            let components = [value.major, value.minor, value.patch].flatMap { $0 }
-            return components.map{ "\($0)" }.joined(separator: ".")
+    public func encode(to encoder: Encoder) throws {
+        let value = self.map { $0.value.description }.joined(separator: ".")
+        var container = encoder.singleValueContainer()
+        try container.encode(value)
+    }
+}
+
+extension Version: Sequence {
+
+    public struct Iterator: IteratorProtocol {
+        private var version: Version?
+
+        init(version: Version?) {
+            self.version = version
         }
-        return nil
+
+        public mutating func next() -> Version? {
+            guard let version = self.version else {
+                return nil
+            }
+
+            if case .subversion(_, let subversion) = version {
+                self.version = subversion
+            } else {
+                self.version = nil
+            }
+
+            return version
+        }
+    }
+
+    public func makeIterator() -> Iterator {
+        return Iterator(version: self)
+    }
+}
+
+extension Version: CustomStringConvertible {
+    public var description: String {
+        return self.map { $0.value.description }.joined(separator: ".")
     }
 }
 
-public struct Version {
-    public let major: Int
-    public let minor: Int?
-    public let patch: Int?
+extension Version: Equatable {
+    public static func ==(left: Version, right: Version) -> Bool {
+        let leftArray = left.map { $0.value }
+        let rightArray = right.map { $0.value }
+        return leftArray == rightArray
+    }
 }
+
+extension Version: Comparable {
+    public static func <(left: Version, right: Version) -> Bool {
+        let leftArray = left.map { $0.value }
```

Comment:
Should there be a `components` property on `Version`?

Reply:
Sure